### PR TITLE
Merge frame-based timeline and full timeline.

### PR DESCRIPTION
**Pull request by kenzieschmoll**
This CL eliminates the notion of "modes" and combines Flutter frame information from the "frame-based" view with the full trace viewer from the "full timeline" view.

Selecting a frame from the frame chart zooms and animates the trace viewer to the events associated with the selected flutter frame.

combo_timeline_small

In order to preserve the html app without spending effort trying to make it compatible with this change, `timeline_controller.dart`, `timeline_model.dart`, and `timeline_processor.dart` were all renamed to `html_*.dart` and forked into the `/fluter` dir. All flutter code in the timeline now imports the timeline files from inside the `/fluter` dir. Some testing files needed to be forked as well.

### CODE REVIEW

File: packages/devtools_app/lib/src/charts/flutter/flame_chart.dart [hunk 1]
```diff
@@ -712,6 +725,8 @@ class FlameChartNode<T> {
   final Color backgroundColor;
   final Color textColor;
   final T data;
+  final bool Function(T) useAlternateBackground;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/charts/flutter/flame_chart.dart]: this should probably be a method rather than a Function.

File: packages/devtools_app/lib/src/charts/flutter/flame_chart.dart [hunk 2]
```diff
@@ -724,6 +739,11 @@ class FlameChartNode<T> {
     @required bool hovered,
     @required double zoom,
   }) {
+    final zoomedWidth = rect.width * zoom;
+    if (zoomedWidth < 0.5) {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/charts/flutter/flame_chart.dart]: Add a TODO: we should probably still show a solid line although perhaps fading it out.
0.5 seems a bit arbitrary for this case. It could probably equally reasonably be 0.2 for example.

File: packages/devtools_app/lib/src/charts/flutter/flame_chart.dart [hunk 3]
```diff
@@ -762,6 +780,22 @@ class FlameChartNode<T> {
...
+    if (useAlternateBackground != null && useAlternateBackground(data)) {
+      return alternateBackgroundColor;
+    }
+    return backgroundColor;
+  }
+
+  Color _textColor(bool selected) {
+    if (selected ||
+        (useAlternateBackground != null && useAlternateBackground(data))) {
+      return Colors.black;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/charts/flutter/flame_chart.dart]: document why this should be black or add a semantically named constant for what this color is.

File: packages/devtools_app/lib/src/charts/flutter/flame_chart.dart [hunk 4]
```diff
@@ -775,40 +809,54 @@ class FlameChartNode<T> {
 
 mixin FlameChartColorMixin {
   int _uiColorOffset = 0;
-  Color nextUiColor() {
+  Color nextUiColor({bool resetOffset = false}) {
+    if (resetOffset) _uiColorOffset = 0;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/charts/flutter/flame_chart.dart]: add curly braces around if statements that don't end with a return.

File: packages/devtools_app/lib/src/flutter/theme.dart [hunk 5]
```diff
@@ -66,6 +66,9 @@ const devtoolsBlue = ColorSwatch<int>(600, {
   400: Color(0xFF13B9FD),
 });
 
+/// A short duration to use for animations.
+const shortDuration = Duration(milliseconds: 50);
```
Comment (jacob314) [inline packages/devtools_app/lib/src/flutter/theme.dart]: maybe give some guidance on when short duration should be used instead of the default duration for others adopting this theme.

File: packages/devtools_app/lib/src/profiler/cpu_profile_model.dart [hunk 6]
```diff
@@ -52,8 +51,8 @@ class CpuProfileData {
     // Each trace event in [subTraceEvents] will have the leaf stack frame id
     // for a cpu sample within [subTimeRange].
     final subTraceEvents = superProfile.stackTraceEvents
-        .where((trace) => subTimeRange
-            .contains(Duration(microseconds: trace[TraceEvent.timestampKey])))
+        .where((trace) =>
+            subTimeRange.contains(Duration(microseconds: trace['ts'])))
```
Comment (jacob314) [inline packages/devtools_app/lib/src/profiler/cpu_profile_model.dart]: why this change? seems like a minor regression to be using`'ts'` instead of `TraceEvent.timestampKey`
Optionally: maybe add an extension method for this?
e.g. 
trace.timeStamp

File: packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart [hunk 7]
```diff
@@ -0,0 +1,350 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart]: can you comment on which parts of this file are new or upload a baseline version that is the code it was forked from?

File: packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart [hunk 8]
```diff
@@ -15,198 +15,52 @@ import '../../geometry.dart';
...
-import '../timeline_model.dart';
+import 'timeline_controller.dart';
+import 'timeline_model.dart';
 
 class TimelineFlameChart extends StatelessWidget {
   @override
   Widget build(BuildContext context) {
     final controller = Controllers.of(context).timeline;
     return LayoutBuilder(builder: (context, constraints) {
       return ValueListenableBuilder(
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart]: having multiple nested ValueListenableBuilders is an anti-pattern. 
Either refactor to use a Stateful widget or create a 
ValueListenablesBuilder widget that takes a couple ValueListenables.

File: packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart [hunk 9]
```diff
@@ -222,25 +76,85 @@ class FullTimelineFlameChart
...
+      // Vertically scroll to the UI event group.
+      final verticalScrollOffset =
+          eventGroupStartXValues[widget.data.eventGroups[TimelineData.uiKey]];
+      await verticalScrollController.animateTo(
+        verticalScrollOffset,
+        duration: shortDuration,
+        curve: defaultCurve,
+      );
+
+      // Zoom the frame into view.
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart]: we need to make sure the animation still makes sense when we get here otherwise things will get janky if a couple animations trigger in succession. same for all other animations within this method.
An easy way to track it would be to make sure _selectedFrame == selectedFrame and bail out otherwise.

File: packages/devtools_app/lib/src/timeline/flutter/timeline_model.dart [hunk 10]
```diff
@@ -0,0 +1,1099 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'dart:collection';
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_model.dart]: update copyright notice or update the diff to include the file this is based on.

File: packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart [hunk 11]
```diff
@@ -0,0 +1,615 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart]: update the diff to show the file this is based on assuming you are just forking a file that wasn't under the Flutter directory.

File: packages/devtools_app/lib/src/timeline/flutter/timeline_screen.dart [hunk 12]
```diff
@@ -89,22 +83,19 @@ class TimelineScreenBodyState extends State<TimelineScreenBody>
     controller.timelineService.updateListeningState(true);
 
     cancel();
-    addAutoDisposeListener(controller.timelineModeNotifier);
-    addAutoDisposeListener(controller.fullTimeline.recordingNotifier, () {
+    addAutoDisposeListener(controller.recordingNotifier, () {
       setState(() {
-        recording = controller.fullTimeline.recordingNotifier.value;
+        recording = controller.recordingNotifier.value;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_screen.dart]: nit: recordingNotifier seems like a bit of an awkward name when used here.
maybe it is just
controller.recording with a type ValueNotifier?

File: packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart [hunk 13]
```diff
@@ -0,0 +1,167 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:vm_service/vm_service.dart' as vm_service;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart]: can you provide the baseline version of the code this is based on?

File: packages/devtools_app/test/flutter/timeline_model_test.dart [hunk 14]
```diff
@@ -0,0 +1,317 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
Comment (jacob314) [inline packages/devtools_app/test/flutter/timeline_model_test.dart]: welcome to 2020 here and elsewhere.
Please also mark tests that should be delete when the dart:html app is removed as part of this CL.

Comment (kenzieschmoll) [reply packages/devtools_app/test/flutter/timeline_model_test.dart]: I left it as 2019 because I copied it from test/timeline_model_test.dart.

File: packages/devtools_app/lib/src/charts/flutter/flame_chart.dart (hunk 2 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/charts/flutter/flame_chart.dart]: added TODO

File: packages/devtools_app/lib/src/charts/flutter/flame_chart.dart (hunk 3 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/charts/flutter/flame_chart.dart]: done

File: packages/devtools_app/lib/src/charts/flutter/flame_chart.dart (hunk 4 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/charts/flutter/flame_chart.dart]: done

File: packages/devtools_app/lib/src/flutter/theme.dart (hunk 5 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/flutter/theme.dart]: done

File: packages/devtools_app/lib/src/profiler/cpu_profile_model.dart (hunk 6 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/profiler/cpu_profile_model.dart]: I did this to remove the timeline_model import. `cpu_profile_model.dart` is used by both the html and flutter implementations of DevTools, and having it import one or the other was causing dependency issues. I'll add a TODO to use `TraceEvent.timestampKey` once we delete the html app.

File: packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart (hunk 8 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart]: done.

File: packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart (hunk 9 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart]: Like so? (ptal)

File: packages/devtools_app/lib/src/timeline/flutter/timeline_model.dart [hunk 15]
```diff
@@ -0,0 +1,1099 @@
...
+
+import 'package:meta/meta.dart';
+
+import '../../profiler/cpu_profile_model.dart';
+import '../../service_manager.dart';
+import '../../trees.dart';
+import '../../utils.dart';
+import 'timeline_controller.dart';
+
+class TimelineData {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_model.dart]: in this file I combined the classes  `FrameBasedTimelineData` and `FullTimelineData` into `TimelineData` (which was previously the abstract superclass of both).

File: packages/devtools_app/lib/src/timeline/flutter/timeline_model.dart (hunk 10 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_model.dart]: updated copyright

File: packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart [hunk 16]
```diff
@@ -0,0 +1,615 @@
...
+/// be able to dump this buffer to a downloadable text file.
+StringBuffer debugFrameTracking = StringBuffer();
+
+const String gpuEventName = 'GPURasterizer::Draw';
+
+const String uiEventName = 'VSYNC';
+
+/// Processor for composing a recorded list of trace events into a timeline of
+/// [AsyncTimelineEvent]s, [SyncTimelineEvent]s, and [TimelineFrame]s.
+class TimelineProcessor {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart]: In this file I combined the logic that was previously split between `FrameBasedTimelineProcessor` and `FullTimelineProcessor`

File: packages/devtools_app/lib/src/timeline/flutter/timeline_screen.dart (hunk 12 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_screen.dart]: updated this var and similar ones in TimelineController

File: packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart (hunk 7 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart]: If you look at the latest commit, you can see the changes against the forked version.

File: packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart (hunk 11 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart]: See latest commit

File: packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart (hunk 13 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart]: see latest commit

File: packages/devtools_app/lib/src/profiler/cpu_profile_model.dart (hunk 6 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/profiler/cpu_profile_model.dart]: That is fine. you could probably also move TraceEvent into a separate library used in both places but it isn't critical. Probably makes sense to move TraceEvent outside of timeline_model at some point anyway as it is aa bit different than the rest of timeline_model.dart

File: packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart [hunk 17]
```diff
@@ -131,6 +103,10 @@ class _TimelineFlameChartState
         curve: defaultCurve,
       );
 
+      // Maybe bail early if multiple frame selections were triggered in
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart]: remove this comment. you are already bailing early enough. the existing vertical animation is already going to get cancelled so you are good. 
what might be a good idea is to before the setState return `if (_selectedFrame == selectedFrame)
`
as no animation is needed in that case. That would avoid the case where you could try to re-trigger an animation while the same animation is already running.

Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart]: Nit for this comment. Make it more perscriptive.
E.g.
` // Bail early if the selection has changed again while the animation was in progress.`

File: packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart (hunk 9 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart]: yep!

File: packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart [hunk 18]
```diff
@@ -29,81 +29,76 @@ const String timelineScreenId = 'timeline';
...
   final cpuProfilerController = CpuProfilerController();
 
   /// Notifies that a timeline event was selected.
-  ValueListenable get selectedTimelineEventNotifier =>
-      _selectedTimelineEventNotifier;
+  ValueListenable get selectedTimelineEvent => _selectedTimelineEventNotifier;
   final _selectedTimelineEventNotifier = ValueNotifier<TimelineEvent>(null);
 
+  /// Notifies that a timeline frame has been selected.
+  ValueListenable get selectedFrame => _selectedFrameNotifier;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart]: maybe we should turn on the no-implicit dynamic lints. I wonder how many other places we have dynamic sneaking in where we don't intend it.
Change these all to
ValueListenable<someType>

File: packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart [hunk 19]
```diff
@@ -29,81 +29,76 @@ const String timelineScreenId = 'timeline';
...
+  final _selectedFrameNotifier = ValueNotifier<TimelineFrame>(null);
+
+  /// Notifies when an empty timeline recording finishes
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart]: Nit. Remove the Notifies from all these comments. the object isn't notifying. It is just holding a value and letting users listen.
E.g. maybe
`/// Notifies that the timeline is currently being recorded.`
-->
`/// Whether the timeline is currently being recorded.`

File: packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart [hunk 20]
```diff
@@ -29,81 +29,76 @@ const String timelineScreenId = 'timeline';
...
+  ValueListenable get emptyRecording => _emptyRecordingNotifier;
+  final _emptyRecordingNotifier = ValueNotifier<bool>(false);
+
+  /// Notifies that the timeline is currently being recorded.
+  ValueListenable get recording => _recordingNotifier;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart]: @pq refactoring microfriction: it would be nice that when you renamed a property, the private field associated with it also renamed.
In this case the property was original `recordingNotifier => _recordingNotifier`
but refactoring the name of `recordingNotifier` didn't rename `_recordingNotifier`

File: packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart [hunk 21]
```diff
@@ -29,81 +29,76 @@ const String timelineScreenId = 'timeline';
...
+  final _recordingNotifier = ValueNotifier<bool>(false);
+
+  /// Notifies that the recorded timeline data is currently being processed.
+  ValueListenable get processing => _processingNotifier;
+  final _processingNotifier = ValueNotifier<bool>(false);
+
+  /// Stream controller that notifies the timeline has been processed.
+  Stream<bool> get onTimelineProcessed => _timelineProcessedController.stream;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart]: Why is this one a Stream while the rest are ValueListenable?

File: packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart [hunk 22]
```diff
@@ -29,81 +29,76 @@ const String timelineScreenId = 'timeline';
...
+  final _timelineProcessedController = StreamController<bool>.broadcast();
+
   /// Stream controller that notifies that offline data was loaded into the
   /// timeline.
   ///
   /// Subscribers to this stream will be responsible for updating the UI for the
   /// new value of [timelineData].
-  final _loadOfflineDataController = StreamController<OfflineData>.broadcast();
+  final _loadOfflineDataController =
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart]: same for this case, consider refactoring all the streams to ValueListenable either with this CL or with a followup.

File: packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart [hunk 23]
```diff
@@ -29,81 +29,76 @@ const String timelineScreenId = 'timeline';
...
+  /// This does not include events that we receive while stopped.
   List<TraceEventWrapper> allTraceEvents = [];
 
-  bool get hasStarted =>
-      frameBasedTimeline.hasStarted && fullTimeline.hasStarted;
-
-  void selectTimelineMode(TimelineMode mode) {
-    _timelineModeNotifier.value = mode;
-  }
+  bool get hasStarted => data != null;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart]: should this be true if there is offline data?

File: packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart [hunk 24]
```diff
@@ -29,81 +29,76 @@ const String timelineScreenId = 'timeline';
...
-      timelineModeNotifier.value == TimelineMode.frameBased
-          ? frameBasedTimeline
-          : fullTimeline;
-
-  FrameBasedTimeline frameBasedTimeline;
-
-  FullTimeline fullTimeline;
-
-  List<TimelineBase> timelines;
+  TimelineData data;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart]: we have both data and OfflineTimelineData.
Document expectations on whether one or both can be non-null.

File: packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart [hunk 25]
```diff
@@ -122,23 +117,95 @@ class TimelineController implements DisposableController {
...
+      frame.gpuEventFlow.writeTraceToBuffer(buf);
+      log(buf.toString());
+    }
+  }
+
+  void addFrame(TimelineFrame frame) {
+    data.frames.add(frame);
+  }
+
+  int vmStartRecordingMicros;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart]: should this be moved with the other fields?

File: packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart [hunk 26]
```diff
@@ -153,17 +220,15 @@ class TimelineController implements DisposableController {
...
-    timeline
-      ..data = offlineData.shallowClone()
-      ..processor.primeThreadIds(
-        uiThreadId: uiThreadId,
-        gpuThreadId: gpuThreadId,
-      );
-    await timeline.processTraceEvents(traceEvents);
-
-    if (timeline.data.cpuProfileData != null) {
+    data = offlineData.shallowClone();
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart]: why are there offlineTimelineData and data members with the same content?
consider removing offlineTimelineData and maybe just storing a bool that the data is offline?

File: packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart [hunk 27]
```diff
@@ -643,7 +98,15 @@ class FullTimelineProcessor extends TimelineProcessor {
...
-  Future<void> processTimeline(List<TraceEventWrapper> traceEvents) async {
+  // TODO(kenz): Remove the [uiThreadId] and [gpuThreadId] once ui/gpu
+  //  distinction changes and frame ids are available in the engine.
+  int uiThreadId;
+
+  int gpuThreadId;
+
+  Future<void> processTimeline(
+      List<TraceEventWrapper> traceEvents, int startRecordingMicros,
+      {bool resetAfterProcessing = true}) async {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart]: missing trailing comma

File: packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart [hunk 28]
```diff
@@ -654,11 +117,21 @@ class FullTimelineProcessor extends TimelineProcessor {
...
-        // Throw out timeline events that do not have a timestamp
-        // (e.g. thread_name events).
-        .where((event) => event.event.timestampMicros != null)
-        .toList())
+    final _traceEvents = (traceEvents.where((event) {
+      // Throw out 'MessageLoop::FlushTasks' events. A single
+      // 'MessageLoop::FlushTasks' event can parent multiple Flutter frame event
+      // sequences and complicates the frame detection logic.
+      final isMessageLoopFlushTasks =
+          event.event.name.contains('MessageLoop::FlushTasks');
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart]: Maybe move all the magic event name constants e.g (`'MessageLoop::FlushTasks'`) we match to be top level fields in this library or a helper library with a prefix. That way you can audit at a glance all the special cases we are handling differently.

File: packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart [hunk 29]
```diff
@@ -739,9 +225,17 @@ class FullTimelineProcessor extends TimelineProcessor {
         case TraceEvent.durationCompletePhase:
           _handleDurationCompleteEvent(eventWrapper);
           break;
+        // TODO(kenz): add additional support for flows.
         case TraceEvent.flowStartPhase:
+          if (eventWrapper.event.name.contains('PipelineItem')) {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart]: PipelineItem is another case we're treating magically. Can we make this pattern more restricted so we have a lower risk of false positives? E.g. what is the range of what the event names look like in practice?

File: packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart [hunk 30]
```diff
@@ -825,17 +318,22 @@ class FullTimelineProcessor extends TimelineProcessor {
   }
 
   void _handleDurationBeginEvent(TraceEventWrapper eventWrapper) {
-    final current = _currentDurationEventNodes[eventWrapper.event.threadId];
+    final current = currentDurationEventNodes[eventWrapper.event.threadId];
     final timelineEvent = SyncTimelineEvent(eventWrapper);
+    if (current == null) {
+      if (eventWrapper.event.name.contains(uiEventName)) {
+        // pending frame . ui event flow = timeline event
+      } else if (eventWrapper.event.name.contains(gpuEventName)) {}
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart]: remove no-op else case and change if to `if (current == null && eventWrapper.event.name.contains(uiEventName))`

File: packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart [hunk 31]
```diff
@@ -825,17 +318,22 @@ class FullTimelineProcessor extends TimelineProcessor {
...
       current.addChild(timelineEvent);
     }
-    _currentDurationEventNodes[eventWrapper.event.threadId] = timelineEvent;
+    currentDurationEventNodes[eventWrapper.event.threadId] = timelineEvent;
   }
 
   void _handleDurationEndEvent(TraceEventWrapper eventWrapper) {
     final TraceEvent event = eventWrapper.event;
-    SyncTimelineEvent current = _currentDurationEventNodes[event.threadId];
+    SyncTimelineEvent current = currentDurationEventNodes[event.threadId];
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart]: add final

File: packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart [hunk 32]
```diff
@@ -972,29 +473,125 @@ class FullTimelineProcessor extends TimelineProcessor {
...
+      if (pendingFrame.pipelineItemTime.start.inMicroseconds ==
+          event.timestampMicros) {
+        pendingFrame.pipelineItemStartTrace = event;
+      }
+      _maybeAddCompletedFrame(pendingFrame);
+    }
+  }
+
+  void _handleFrameEndEvent(TraceEvent event) async {
+    if (event.id != null) {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart]: reduce nesting by writing as
`if (event.id == null) return;`

File: packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart [hunk 33]
```diff
@@ -972,29 +473,125 @@ class FullTimelineProcessor extends TimelineProcessor {
...
+        pendingFrame.pipelineItemEndTrace = event;
+      }
+      _maybeAddCompletedFrame(pendingFrame);
+    }
+  }
+
+  /// Add event to an available frame in [pendingFrames] if we can.
+  void _maybeAddFlutterFrameEvent(SyncTimelineEvent event) {
+    if (!event.isUiEventFlow && !event.isGpuEventFlow) {
+      // We do not care about events that are neither the main flow of UI
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart]: add a TODO and bug to figure out how to still show these events.

File: packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart (hunk 13 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart]: thanks!

File: packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart [hunk 34]
```diff
@@ -23,6 +22,10 @@ class TimelineService {
 
   final profilerService = CpuProfilerService();
 
+  Future<vm_service.Timestamp> vmTimelineMicros() async {
+    return await serviceManager.service.getVMTimelineMicros();
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart]: follow up cl: rename 
serviceManager.service.getVMTimelineMicros to serviceManager.service.vmTimelineMicros

File: packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart [hunk 35]
```diff
@@ -31,44 +34,29 @@ class TimelineService {
       _handleConnectionStart(serviceManager.service);
     }
     serviceManager.onConnectionClosed.listen(_handleConnectionStop);
+
+    timelineController.recording.addListener(() async {
+      await updateListeningState(true);
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart]: what purpose is this `async` and `await` serving? 
`() => updateListeningState(true)`
should have the same impact.

File: packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart [hunk 36]
```diff
@@ -80,17 +68,19 @@ class TimelineService {
 
   Future<void> startTimeline() async {
     if (await serviceManager.connectedApp.isAnyFlutterApp) {
-      timelineController.frameBasedTimeline.data = FrameBasedTimelineData(
+      timelineController.data = TimelineData(
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart]: nit: 
slightly terser:
```dart
timelineController.data = serviceManager.connectedApp.isAnyFlutterApp ? TimelineData(displayRefreshRate: await serviceManager.getDisplayRefreshRate())  : TimelineData();
```

File: packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart [hunk 37]
```diff
@@ -80,17 +68,19 @@ class TimelineService {
...
     }
-    timelineController.fullTimeline.data = FullTimelineData();
 
     await serviceManager.serviceAvailable.future;
     await allowedError(serviceManager.service
         .setVMTimelineFlags(<String>['GC', 'Dart', 'Embedder']));
     await allowedError(serviceManager.service.clearVMTimeline());
 
-    final Timeline timeline = await serviceManager.service.getVMTimeline();
+    final vm_service.Timeline timeline =
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart]: nit: is this type on the lhs useful?

File: packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart (hunk 29 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart]: for a flutter app, PipelineItem events are flow events that signal frame start and frame end. I made this a const and added a comment to explain.

File: packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart (hunk 30 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart]: this slipped into the commit by accident. removed the if else clause

File: packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart (hunk 32 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart]: done

File: packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart (hunk 33 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_processor.dart]: we do still show these events, we just don't assign them to a Flutter frame.

File: packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart (hunk 34 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart]: this would require changing the vm_service API. We are just overriding the method in vm_service_wrapper

File: packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart (hunk 35 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart]: done

File: packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart (hunk 36 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart]: done

File: packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart (hunk 37 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart]: removed. Also removed prefix -no longer needed

File: packages/devtools_app/lib/src/charts/flutter/flame_chart.dart (hunk 1 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/charts/flutter/flame_chart.dart]: discussed offline. Added a related todo to FlameChartNode class

File: packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart (hunk 21 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart]: @terrylucas FYI

File: packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart [hunk 38]
```diff
@@ -90,10 +90,16 @@ class TimelineFlameChartState
   void _handleSelectedFrame() async {
     final TimelineFrame selectedFrame = _timelineController.selectedFrame.value;
     if (selectedFrame != null) {
+      // Bail early if the selection has not changed.
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart]: nit: no real need to document this line. it is just the standard pattern of skipping setState when the state has not  changed.

File: packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart [hunk 39]
```diff
@@ -90,10 +90,16 @@ class TimelineFlameChartState
...
+      if (selectedFrame == _selectedFrame) return;
+
       setState(() {
         _selectedFrame = selectedFrame;
       });
 
+      // TODO(kenz): consider using jumpTo for some of these animations to bump
+      // performance.
```
Comment (jacob314) [inline packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart]: nit: "bump performance" --> "improve performance"

File: packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart (hunk 34 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_service.dart]: oh right.

File: packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart (hunk 23 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart]: no, this value reflects whether the Timeline has been started from timeline_service.dart. `data` is initialized there (where we also set the timeline recorders like 'Dart', 'GC', etc.). I will add a comment.

File: packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart (hunk 25 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart]: done

File: packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart (hunk 26 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_controller.dart]: we need to store both so that we can revert to the original timeline data if event selections and frame selections while offline modify `data`.

File: packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart (hunk 39 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart]: done
